src/workerd/api/sockets.c++

```diff
@@ -60,122 +69,17 @@ jsg::Ref<Socket> connectImpl(
-    (*stream)->getsockopt(level, option, value, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::setsockopt(int level, int option, const void* value, uint length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->setsockopt(level, option, value, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getsockname(struct sockaddr* addr, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getsockname(addr, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getpeername(struct sockaddr* addr, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getpeername(addr, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-kj::Promise<size_t> PipelinedAsyncIoStream::read(void* buffer, size_t minBytes, size_t maxBytes) {
-  return thenOrRunNow<size_t>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->read(buffer, minBytes, maxBytes);
-  });
-}
-
-kj::Promise<size_t> PipelinedAsyncIoStream::tryRead(void* buffer, size_t minBytes, size_t maxBytes) {
-  return thenOrRunNow<size_t>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->tryRead(buffer, minBytes, maxBytes);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::write(const void* buffer, size_t size) {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->write(buffer, size);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->write(pieces);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::whenWriteDisconnected() {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->whenWriteDisconnected();
-  });
+  // Forcibly close the readable/writable streams.
+  auto cancelPromise = readable->getController().cancel(js, nullptr);
+  auto abortPromise = writable->getController().abort(js, nullptr);
+  // The below is effectively `Promise.all(cancelPromise, abortPromise)`
+  return cancelPromise.then(js, [abortPromise = kj::mv(abortPromise), this](jsg::Lock& js) mutable {
+    return abortPromise.then(js, [this](jsg::Lock& js) {
+      resolveFulfiller(nullptr);
+      return js.resolvedPromise();
+    }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
+  }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
```

Comment:
So it looks like any exceptions caught from `cancel()` or `abort()` propagate to the `closed` promise, but do NOT propagate to the result of `close()` itself.

Is that what we want? I'm not really sure when `cancel()` or `abort()` would throw, but it would seem to me that the exception should propagate out of `close()`. Meanwhile it's not clear to me whether the `closed` promise should signal errors or just signal that the socket is now closed.

Reply:
Well, I'd allow the error to propagate to the `closed` promise but I would explicitly mark it handled using `markAsHandled` so that if someone isn't paying attention it's perfectly ok.

Reply:
> Meanwhile it's not clear to me whether the closed promise should signal errors or just signal that the socket is now closed.

It's not clear to me either. Though I trust James' opinion here.

This is another case that I think we need to discuss in more depth. I want us to do so on this issue and many others, after discussion we may end up with another decision (especially if we decide to conform to [other specs](https://wicg.github.io/direct-sockets/)) so let's not worry too much about making it perfect in this PR.